internal/cloud/instanceprotector/instance-protector.go

```diff
@@ -0,0 +1,187 @@
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"github.com/sirupsen/logrus"
+)
+
+type InstanceProtector struct {
+	protect   chan struct{}
+	unprotect chan struct{}
+}
+
+// NewInstanceProtector returns an instance protector that can
+// protect up to @size jobs in parallel. Beyond that limit
+// calls to Protect() starts to block.
+func NewInstanceProtector(size uint64) *InstanceProtector {
+	ip := &InstanceProtector{}
+	ip.protect = make(chan struct{}, size)
+	ip.unprotect = make(chan struct{}, size)
+	return ip
+}
+
+func (ip *InstanceProtector) Protect() {
+	if ip != nil {
+		ip.protect <- struct{}{}
+	}
+}
+
+func (ip *InstanceProtector) Unprotect() {
+	if ip != nil {
+		ip.unprotect <- struct{}{}
+	}
+}
+
+func (ip *InstanceProtector) Start(ctx context.Context) {
+	// Perform alternate calls to protect / unprotect until
+	// ctx is cancelled.
+	//
+	// Collate matching protect/unprotect tokens when possible
+	// to avoid uneccesary calls (which can be slow), this should
+	// minimize the risk of queues of tokens building up which in
+	// the worst case could block new work being scheduled, or at
+	// least cause us to unprotect when we should still be protected
+	// and vice versa.
+	//
+	// The main reason for doing this async, is that the calls can
+	// be very slow and cause unneccesary blocking. There is already
+	// a race between a job being dequeued and the instance being
+	// protected, and moreover the protection is not always respected
+	// by the cloud provider, so we do not lose much by being async.
+	for {
+	protect:
+		for {
+			// Pop off alternating protect / unprotect tokens until we receive
+			// a protect token without a matching unprotect one for ten seconds.
+			// At that point, protect the instance and break from this loop.
+			//
+			// We wait for ten seconds to avoid a busy loop, and allow more
+			// collation. At worst this means we have to redo ten seconds of
+			// work in case the instance is killed early (but that can anyway
+			// happen for other reasons).
+
+			var timeout <-chan time.Time
+			select {
+			case <-ip.protect:
+				// received the initial protect request, do not act on it
+				// for at least ten seconds
+				timeout = time.After(10 * time.Second)
+			case <-ctx.Done():
+				return
+			}
+			for {
+				// Drain the channels of already scheduled protect / unprotect pairs
+				select {
+				case <-ip.unprotect:
+				case <-timeout:
+					// There has been more protect than unprotect requests for
+					// the duration of the timeout. Ready to set protection,
+					// break out of the loop.
+					break protect
+				}
+				select {
+				case <-ip.protect:
+				default:
+					// All protect and unprotect calls have been matched up before
+					// the timeout expired and no protect requests are pending. Discard
+					// the request to protect the instance and start over.
+					continue protect
+				}
+			}
+		}
+		setProtection(true)
+		defer setProtection(false)
```

Comment:
This shouldn't be here. `defer` is scoped to a function, not to a block. `setProtection` should be therefore called before `break unprotect` after line 118.

Reply:
Thanks! I think I might add a helper function to get the right scope. I was trying to make it obvious what's going on, still not quite there ;) I'll also add a test for this function.

Reply:
Fixed.